tests/core/middleware/test_transaction_signing.py

```diff
@@ -0,0 +1,188 @@
+    w3_base.middleware_stack.add(result_generator_middleware)
+    return w3_base
+
+
+@pytest.mark.parametrize(
+    'method,expected',
+    (
+        ('eth_sendTransaction', 'eth_sendRawTransaction'),
+        ('eth_call', NotImplementedError),
+    )
+)
+def test_sign_and_send_raw_middleware(w3_dummy, w3, method, expected, key_object):
+    w3_dummy.middleware_stack.add(construct_sign_and_send_raw_middleware(key_object))
+    account = to_account(key_object)
+
+    if isinstance(expected, type) and issubclass(expected, Exception):
+        with pytest.raises(expected):
+            w3_dummy.manager.request_blocking(
+                method,
+                [{
+                    'to': '0x7E5F4552091A69125d5DfCb7b8C2659029395Bdf',
+                    'from': account.address,
+                    'gas': 21000,
+                    'gasPrice': 0,
+                    'value': 1,
+                    'nonce': 0
+                }])
+    else:
+        actual = w3_dummy.manager.request_blocking(
+            method,
+            [{
+                'to': '0x7E5F4552091A69125d5DfCb7b8C2659029395Bdf',
+                'from': account.address,
+                'gas': 21000,
+                'gasPrice': 0,
+                'value': 1,
+                'nonce': 0
+            }])
+        raw_txn = actual[1][0]
+        actual_method = actual[0]
+        assert actual_method == expected
+        assert isinstance(raw_txn, bytes)
+
+
+@pytest.fixture(scope='session')
+def w3():
+    return Web3(EthereumTesterProvider())
+
+
+@pytest.fixture(
+    scope='session',
+    params=[
+        eth_keys.keys.PrivateKey,
+        eth_account.Account.privateKeyToAccount,
+        HexBytes,
+    ])
+def key_object(request, private_key):
+    return request.param(private_key)
+
+
+def test_to_account(key_object):
+    account = to_account(key_object)
+    assert isinstance(account, eth_account.local.LocalAccount)
+
+
+def test_to_account_type_error(w3):
+    with pytest.raises(TypeError):
+        to_account(1234567890)
+
+
+@pytest.fixture(scope='session')
+def fund_account(w3, private_key):
+    # fund local account
+    account = to_account(private_key)
+    tx_value = w3.toWei(10, 'ether')
+    w3.eth.sendTransaction({
+        'to': account.address,
+        'from': w3.eth.accounts[0],
+        'gas': 21000,
+        'value': tx_value})
+    assert w3.eth.getBalance(account.address) == tx_value
+
+
+def test_signed_transaction_with_set_gas(w3, key_object, fund_account):
+    account = to_account(key_object)
+    w3.middleware_stack.add(construct_sign_and_send_raw_middleware(key_object))
+    start_balance = w3.eth.getBalance(account.address)
+    w3.eth.sendTransaction({
+        'to': w3.eth.accounts[0],
+        'from': account.address,
+        'gas': 21000,
+        'gasPrice': 0,
+        'value': 1})
+    assert w3.eth.getBalance(account.address) <= start_balance - 1
+
+
+def test_signed_transaction_unset_gas(w3, key_object, fund_account):
```

Comment:
Looks like this and the one above it could be compressed into one with parametrize.